### PR TITLE
Fix unhandled exceptions in ConsensusRuleEngine

### PR DESCRIPTION
**Pull request by fassadlr**
Currently if we encounter an unhandled exception in the `ConsensusRuleEngine`, it is not handled and as such ConsensusManager does not know what to do with the peer it received the invalid block from.

An example is being connected to a Stratis-X node that was upgraded and after the upgrade activation point we receive a block which is now not valid anymore and we cant load it from the stake store, an exception is thrown which is not handled. This causes the peer to still be connected and as such the node fails to advance.

I am not sure whether we should do this as now ALL unhandled exceptions in the `ConsensusRuleEngine` will now cause the peer it got the block from, to be banned and disconnected?

### CODE REVIEW

File: src/Stratis.Bitcoin/Consensus/ConsensusRuleEngine.cs [hunk 1]
```diff
@@ -234,6 +233,10 @@ private async Task ExecuteRulesAsync(IEnumerable<AsyncConsensusRule> asyncRules,
             {
                 ruleContext.ValidationContext.Error = ex.ConsensusError;
             }
+            catch (Exception ex)
+            {
+                ruleContext.ValidationContext.Error = new ConsensusError("unhandled-exception", ex.Message);
```
Comment (ferdeen) [inline src/Stratis.Bitcoin/Consensus/ConsensusRuleEngine.cs]: Could you do `ex.ToString()` here ? As there's an override in `ConsensusError` which includes the error code, as well as the message.

Comment (fassadlr) [reply src/Stratis.Bitcoin/Consensus/ConsensusRuleEngine.cs]: Done :)

**Comment by fassadlr** [comment]
Fixes #2153

File: src/Stratis.Bitcoin/Consensus/ConsensusRuleEngine.cs [hunk 2]
```diff
@@ -234,6 +233,10 @@ private async Task ExecuteRulesAsync(IEnumerable<AsyncConsensusRule> asyncRules,
             {
                 ruleContext.ValidationContext.Error = ex.ConsensusError;
             }
+            catch (Exception ex)
```
Comment (dangershony) [inline src/Stratis.Bitcoin/Consensus/ConsensusRuleEngine.cs]: Probably not the correct solution, but also I don't think this is the problem to the bug you got.
I would say we need to catch such exception at the validators and handles there

**Comment by dangershony** [comment]
>An example is being connected to a Stratis-X node that was upgraded and after the upgrade activation point we receive a block which is now not valid anymore and we cant load it from the stake store  

Is it confirmed this is the problem?
The description does not make sense so much if we dont have the stake info of the current block only of the previous one, which is what is failing to load.

**Comment by dangershony** [comment]
I think the problem most likely is that we don't flush stake chain and coinview in the same time.

**Comment by fassadlr** [comment]
Guys for me the issue is here in `OnPartialValidationSucceededAsync`. If  `connectBlocksResult` result contains an error why do we still continue after the `if` and further onto `StartPartialValidation`?

```
           if (connectBlocksResult != null)
            {
                if (connectBlocksResult.PeersToBan != null)
                {
                    var peersToBan = new List<INetworkPeer>();

                    lock (this.peerLock)
                    {
                        foreach (int peerId in connectBlocksResult.PeersToBan)
                        {
                            if (this.peersByPeerId.TryGetValue(peerId, out INetworkPeer peer))
                                peersToBan.Add(peer);
                        }
                    }

                    this.logger.LogTrace("{0} peers will be banned.", peersToBan.Count);

                    foreach (INetworkPeer peer in peersToBan)
                        this.peerBanning.BanAndDisconnectPeer(peer.PeerEndPoint, connectBlocksResult.BanDurationSeconds, connectBlocksResult.BanReason);
                }

                if (connectBlocksResult.ConsensusTipChanged)
                    await this.NotifyBehaviorsOnConsensusTipChangedAsync().ConfigureAwait(false);

                lock (this.peerLock)
                {
                    this.ProcessDownloadQueueLocked();
                }
            }

            if (chainedHeaderBlocksToValidate != null)
            {
                this.logger.LogTrace("Partial validation of {0} block will be started.", chainedHeaderBlocksToValidate.Count);

                // Start validating all next blocks that come after the current block,
                // all headers in this list have the blocks present in the header.
                foreach (ChainedHeaderBlock toValidate in chainedHeaderBlocksToValidate)
                    this.partialValidator.StartPartialValidation(toValidate.ChainedHeader, toValidate.Block, this.OnPartialValidationCompletedCallbackAsync);
            }
```

**Comment by Aprogiena** [comment]
Yes, this actually looks like the real problem. So please discard the changes in this PR and create a new one that will fix that cause

**Comment by noescape00** [comment]
@fassadlr would you like to fix it or should I?

**Comment by fassadlr** [comment]
@noescape00 I will do the fix , thanks 👍

**Pull request closed without merging**